sdk/src/azure/core/az_hfsm_pipeline.c

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <azure/core/internal/az_hfsm_internal.h>
+#include <azure/core/internal/az_hfsm_pipeline_internal.h>
+#include <azure/core/az_platform.h>
+#include <azure/core/internal/az_precondition_internal.h>
+#include <azure/core/internal/az_result_internal.h>
+
+#include <azure/core/_az_cfg.h>
+
+AZ_NODISCARD az_result
+az_hfsm_pipeline_init(az_hfsm_pipeline* pipeline, az_hfsm_policy* outbound, az_hfsm_policy* inbound)
+{
+  pipeline->_internal.outbound_handler = outbound;
+  pipeline->_internal.inbound_handler = inbound;
+#ifndef TRANSPORT_MQTT_SYNC
+  return az_platform_mutex_init(&pipeline->_internal.mutex);
+#else
+  return AZ_OK;
+#endif
+}
+
+AZ_NODISCARD az_result
+az_hfsm_pipeline_post_outbound_event(az_hfsm_pipeline* pipeline, az_hfsm_event event)
+{
+  az_result ret;
+
+#ifndef TRANSPORT_MQTT_SYNC
```

Comment:
We should add clear documentation that if the transport mode is synchronous, mutexes are not used and the application must ensure exclusion of all operations associated with the object (i.e. pipeline in this case).

Reply:
We should add documentation about thread safety within HFSM as well.